### PR TITLE
Fix Create PPU Cache, Implement SELF precompilation, Savestate fix 1 out of 3 

### LINKED ISSUE #12320: PPU LLVM Precomp not working properly if MGS4 is updated to 2.00
# Quick summary

PPU LLVM Precompilation doesn't work correctly if the game is updated to ver 2.00

## Details

PPU LLVM Precompilation works correctly in the disc version (v1.00), this means that after going through the whole process you won't compile in loading screens, but if it's updated to v2.00 and precompiled, you'll have to compile modules in each loading screen. Can't be reproduced in 25th Anniversary version or Digital version (as both already come as 2.00 and they can't be updated).

Before finding out what it seems to be an issue exclusive to disc versions (which contain the 1.00 data and can be updated) I tried older versions of the emulator in case it was a regression to no avail (will attach some of the logs).

#### Please provide your system configuration:
* OS: Windows 10 Pro
* CPU: Ryzen 7 5800X
* GPU: Nvidia RTX 2060
* Driver version: 516.40.0.0
* RPCS3 version: happens on any version.

#### Please include.
* Anything else you deem to be important

Sadly I don't have 25th Anniversary and haven't dumped my digital version of the game, had 2 friends test for me. Luckily it's easy to replicate but kind of time-consuming. Someone else in the discord did it too and also noticed the SELF files weren't being decrypted, but I don't know if it's related.

[RPCS3 v0.0.15-12071-5ca63816.log](https://github.com/RPCS3/rpcs3/files/9067012/RPCS3.v0.0.15-12071-5ca63816.log)
[RPCS3 v0.19-13069-0ab36ef4.log](https://github.com/RPCS3/rpcs3/files/9067015/RPCS3.v0.19-13069-0ab36ef4.log)
[RPCS3 vv0.0.18-12860-02e1c498.log](https://github.com/RPCS3/rpcs3/files/9067016/RPCS3.vv0.0.18-12860-02e1c498.log)
[RPCS3 vv0.0.23-13857-3840015e.log](https://github.com/RPCS3/rpcs3/files/9067011/RPCS3.vv0.0.23-13857-3840015e.log)

### LINKED ISSUE #14080: "Create PPU Cache" bad folder name
### Quick summary

Cache created from "Create PPU Cache" is invalid.

### Details

The cache folder created from menu option "Create PPU Cache" is missing the "EBOOT.BIN" suffix resulting in cache being recreated when launching the game.

Cache created from "Create PPU Cache": `ppu-5MZzwspwMAdmdjJT0CFxY06q85Ct-`
Cache created when launching the game: `ppu-5MZzwspwMAdmdjJT0CFxY06q85Ct-EBOOT.BIN`

The problem is present on all games.

### Attach a log file

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11848754/RPCS3.log)

### Attach capture files for visual issues

_No response_

### System configuration

RPCS3 0.0.28-15225-715e3856 Alpha | master

### Other details

_No response_

### PR DESCRIPTION
**Pull request by elad335**
* Fix Create PPU Cache, allow it to be usable again.
* Implement PPU SELF precompilation, preventing compilation ingame with many game collections, RDR, GT6 and more.
* Make bad game patches not able to crash the emulator, simply ignore them (bad addresses resulted in segfaults)
* Add one fix of savestate.

Fixes #12320
Fixes #13534

### CODE REVIEW

**Comment by Darkhost1999** [comment]
[Creating Cache.log](https://github.com/RPCS3/rpcs3/files/11857963/RPCS3.log)
[Booting Game.log](https://github.com/RPCS3/rpcs3/files/11857959/RPCS3.log)
While https://github.com/RPCS3/rpcs3/issues/14080 is resolved by this PR. 🏆 

And https://github.com/RPCS3/rpcs3/issues/14024 resolved by https://github.com/RPCS3/rpcs3/pull/14034
There are still issues with the create PPU cache option

[Netflix [NPUP00030].log](https://github.com/RPCS3/rpcs3/files/11857970/RPCS3.log)
The 2 sprx and the .self fail to decrypt

[Call of Duty Ghosts [BLUS31270].log](https://github.com/RPCS3/rpcs3/files/11857977/RPCS3.log) and many more
The Eboot compiles whilst the other .self files say (prx: Invalid ELF type, ovl: Invalid ELF type)

Meanwhile this is the more I was excited about in the other PR. Yes that is 1.10 compiling
image
The other was seeing AC Revelations (Update) compiling when they hadn't before.

TLDR: This issue solves the goal it was intended to solve. The Function isn't perfect.
And I now tested booting the games after creating cache not just creating the cache.

**Comment by Darkhost1999** [comment]
The remaining "issues" with the create PPU cache option I believe are and have just been unimplemented and not result of a PR

**Comment by elad335** [comment]
@Darkhost1999 Retest with latest build.

**Comment by Darkhost1999** [comment]
·F 0:19:35.849168 SIG: Thread terminated due to fatal error: Verification failed
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\System.cpp:1396[:6], in function Load) (error=0x2)
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11859933/RPCS3.log)
You're still pushing changes and so idk if this is useful anymore. lmk when you're ready.

**Comment by elad335** [comment]
Retry with latest when it builds.
Also it would be nice if someone can benchmark MGS4 precompilation and firmware compilation speeds before and after.

**Comment by solarmystic** [comment]
> Retry with latest when it builds. Also it would be nice if someone can benchmark MGS4 precompilation and firmware compilation speeds before and after. Even with the cache in place (should be faster even to just link modules).

@elad335 Using the latest test build, with existing PPU cache in place from latest master, MGS4 crashes and takes down rpcs3 after pressing X on this screen. 

image

Crashlog:- 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11860042/RPCS3.log)

**Comment by elad335** [comment]
Retest.

**Comment by Darkhost1999** [comment]
```
·F 1:48:53.592669 SIG: Thread terminated due to fatal error: Assertion Failed! Vulkan API call failed with unrecoverable error: Initialization failed (VK_ERROR_INITIALIZATION_FAILED)
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\vkutils\device.cpp:687[:3], in function create) (error=0xcb)
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\vkutils\shared.cpp:114[:24], in function die_with_error) (error=0xcb)
```
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11860313/RPCS3.zip)
I don't know if this crash is result of this PR but it forced me to stop testing for a moment.
image

There are still some self files saying SELF: Failed to decrypt SELF metadata info! 
For example: PPU: Failed to precompile 'C:/Games/RPCS3/games//Call of Duty Black Ops [BLUS30591]/PS3_GAME/USRDIR/t5mp_ps3f.self' (prx: Invalid ELF type, ovl: Invalid ELF type): Attempting tratment as executable file

This has me a little bit confused https://github.com/RPCS3/rpcs3/pull/14090/commits/9d8d2a397cb6590bfe5602d0ffb85ffd3ba55caf
> PPU LLVM: Precompile all disc game data including PRX etc 

Are SPRX files being implemented for precompilation as well?
image
Cause I don't think SPRX is working.
Self is working more than before and that is good. Booting games normally will finish what is missing when needed.

I'm going back to try and finish compiling but after obtaining new build

**Comment by elad335** [comment]
"Attempting treatment as executable file" is not an error it means it works.

**Comment by ItsALH** [comment]
The whole emulator crashes after booting precompiled MGS4 v2.00 (doesn't even reach the press X screen)

[4e0622bd.log](https://github.com/RPCS3/rpcs3/files/11860399/RPCS3.log)
[0c94f24.log](https://github.com/RPCS3/rpcs3/files/11860426/RPCS3.log)

**Comment by Darkhost1999** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11860590/RPCS3.log.gz)
Part 2 caches are created successfully. Now to boot and play some games.

**Comment by Megamouse** [comment]
Please detach the bugfixes from the enhancements (which should each be a seperate PR anyway)

**Comment by Ordinary205** [comment]
I decided to test it on Watch Dogs, and start the timer for how long it takes to compile/link PPU modules.

Master (15240 Build):  
1 minute 31 seconds

PR:
2 minutes 35 seconds


[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11866385/RPCS3.log.gz)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11866273/RPCS3.log.gz)

**Comment by Ordinary205** [comment]
Here's the second results:

Master (15240 Build):
1 minute 31 seconds

PR (14cfcee3 Build):
1 minute 31 seconds

PR log:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11867560/RPCS3.log.gz)

No differences of what I can see.

**Comment by ItsALH** [comment]
MGS4 v2.00 still compiles in loading screens despite being precompiled (I'll try to keep up with the commits, but this game takes over an hour)

[PPU Cache precomp.log](https://github.com/RPCS3/rpcs3/files/11870056/RPCS3.log)
[run after precompile.log](https://github.com/RPCS3/rpcs3/files/11870141/RPCS3.log)

**Comment by elad335** [comment]
Ah because MGS4.SELF is encrypted with special key, rip then, those keys need to be obtained when the game is running in runtime.

**Comment by ItsALH** [comment]
> Ah because MGS4.SELF is encrypted with special key, rip then, those keys need to be obtained when the game is running in runtime.

Well, that is unfortunate 😢

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 1]
```diff
@@ -3229,15 +3222,13 @@ extern void ppu_precompile(std::vector<std::string>& dir_queue, std::vector<ppu_
...
 					obj.clear(), src.close(); // Clear decrypted file and elf object memory
 
-					ppu_initialize(*ovlm);
-
-					for (auto& seg : ovlm->segs)
 					{
-						vm::dealloc(seg.addr);
+						// Does not really require this lock, this is done for performance reasons.
+						// Seems like too many created threads is hard for Windows to manage efficiently with many CPU threads.
+						std::lock_guard lock(ovl_mutex);
```
[the remark this answers is not in the document]
Comment (elad335) [reply rpcs3/Emu/Cell/PPUThread.cpp]: Fixed.
